server/src/internalClusterTest/java/org/elasticsearch/cluster/coordination/NodeJoiningIT.java

```diff
@@ -0,0 +1,257 @@
+        // Determine upfront who we want the next master to be
+        final var newMasterNodeName = randomValueOtherThan(originalMasterNodeName, () -> randomFrom(nodeNames));
+
+        // Ensure the logging is as expected
+        try (var mockLog = MockLog.capture(NodeJoinExecutor.class)) {
+
+            // Sets MockTransportService behaviour
+            for (final var transportService : internalCluster().getInstances(TransportService.class)) {
+                final var mockTransportService = asInstanceOf(MockTransportService.class, transportService);
+
+                if (mockTransportService.getLocalNode().getName().equals(newMasterNodeName) == false) {
+                    List<String> listOfActionsToBlock = List.of(
+                        // This forces the current master node to fail
+                        PublicationTransportHandler.PUBLISH_STATE_ACTION_NAME,
+                        // This disables pre-voting on all nodes except the new master, forcing it to win the election
+                        StatefulPreVoteCollector.REQUEST_PRE_VOTE_ACTION_NAME
+                    );
+                    blockActionNameOnMockTransportService(mockTransportService, listOfActionsToBlock);
+                }
+            }
+
+            // We do not expect to see a WARN log about a node disconnecting (#ES-11449)
+            addJoiningNodeDisconnectedWarnLogFalseExpectation(mockLog);
+
+            // We haven't changed master nodes yet
+            assertEquals(originalMasterNodeName, internalCluster().getMasterName());
+
+            // Sends a node join request to the original master node. This will fail, and cause a master failover
+            // startDataOnlyNode waits for the new node to be added, and this can only occur after a re-election
+            String newNodeName = internalCluster().startDataOnlyNode();
+            assertNotEquals(originalMasterNodeName, internalCluster().getMasterName());
+            logger.info("New master is elected");
+
+            // Assert all nodes have accepted N into their cluster state
+            assertNewNodeIsInAllClusterStates(newNodeName);
+
+            mockLog.assertAllExpectationsMatched();
+
+            // Assert the new data node was added
+            DiscoveryNodes discoveryNodes = internalCluster().clusterService().state().nodes();
+            assertEquals(numberOfNodesOriginallyInCluster + 1, discoveryNodes.getSize());
+            assertTrue(getListOfDataNodeNamesFromCluster(newMasterNodeName).contains(newNodeName));
+        }
+    }
+
+    /*
+        In this scenario, node N attempts to join a cluster, there is an election and the original master is re-elected.
+        Node N should join the cluster, but it should not be disconnected (#ES-11449)
+     */
+    @TestLogging(reason = "test includes assertions about logging", value = "org.elasticsearch.cluster.coordination:INFO")
+    public void testNodeTriesToJoinClusterAndThenSameMasterIsElected() {
```

Comment:
Btw this test is a really good discriminator of the improvement, it failed for me 9 out of 10 times I ran it having reverted the behaviour change in `Coordinator`, and passed 10 out of 10 with the change in place. Great stuff.